RainierOne_launch_code_v1.0.0.ino

```diff
@@ -0,0 +1,237 @@
+
+#define NUMDATAFIELDS   21      // Number of data fields for each packet.
+#define MAXBUFSIZE      200     // Maximum buffer size for serial packet.
+#define MAXFIELDSIZE    20      // Maximum size of any data field in the serial packet.
+
+#define SUCCESS         0       // Success return code.
+#define ERROR           -1      // Error return code.
+
+
+// Struct to contain all of the NRNSP flight data in one object.
+typedef struct NRdata
+{
+  char flight_state;
+  double exptime;
+  double altitude;
+  double velocity[3];
+  double acceleration[3];
+  double attitude[3];
+  double angular_velocity[3];
+  bool liftoff_warn;
+  bool rcs_warn;
+  bool escape_warn;
+  bool chute_warn;
+  bool landing_warn;
+  bool fault_warn;
+} NRdata;
+
+
+int parse_serial_packet(const char* buf, NRdata* flight_data);
+
+
+void setup()
+{
+  Serial.begin(115200);     // Set baud rate to 115200 (Default serial configureation is 8N1).
+  Serial.setTimeout(20);    // Set timeout to 20ms (It may take up to 17ms for all of the data to
+                            // transfer from the NRNSP, this ensures that enough time has passed
+                            // to allow for a complete transfer before timing out).
+  
+////////////////////////////////////////////////////////////////////////////////////////////////////
+// Set up LED light source
+////////////////////////////////////////////////////////////////////////////////////////////////////
+  pinMode(2, OUTPUT);
+  digitalWrite(2, LOW);
+////////////////////////////////////////////////////////////////////////////////////////////////////
+  
+}
+
+
+void loop()
+{
+  char buffer[MAXBUFSIZE + 1] = {0};      // Buffer for receiving serial packets.
+  int res;                                // Value for storing results of function calls.
+  NRdata flight_info;                     // Struct for holding current flight data.
+
+  // Initialize the struct and all its data to 0.
+  memset(&flight_info, 0, sizeof(NRdata));
+
+  // Loop forever on receiving data packets.
+  while (1)
+  {
+    // Wait for serial input (Delay 1ms between polling serial port).
+    while(!Serial.available())
+    {
+      delay(1);
+    }
+
+    // Read in the serial data up to the maximum serial packet size.
+    res = Serial.readBytes(buffer, MAXBUFSIZE);
+
+    // If no bytes are read then go back to waiting.
+    if (res == 0)
+    {
+      continue;
+    }
+
+    // Null terminate the buffer (Possibly unnecessary).
+    buffer[res] = 0;
+
+    // Update the current flight info with the new data.
+    res = parse_serial_packet(buffer, &flight_info);
+
+    // If the program failed to parse the new data then go back to waiting.
+    if (res != SUCCESS)
+    {
+      continue;
+    }
+    
+    
+////////////////////////////////////////////////////////////////////////////////////////////////////
+// Turn on light & camera when power is on (this part is a loop where all on/off founctions should 
```

Comment:
Looks fine for now - we don't know what pins everything will use yet. I'm not quite sure what lights we are using yet-some led strips require pwm drivers. We will probably need to experiment a bit to see what works best.